src/openssl/digest/digest.cr

```diff
@@ -0,0 +1,88 @@
+require "../lib_crypto"
+require "./digest_base"
+
+module OpenSSL
+  class Digest
+    class DigestError < OpenSSL::Error; end
+
+    include DigestBase
+
+    getter name
+
+    def initialize(@name, @ctx : LibCrypto::EVP_MD_CTX)
+      raise DigestError.new("Invalid EVP_MD_CTX") unless @ctx
+    end
+
+    protected def self.create_evp_mt_ctx(name)
+      md = LibCrypto.evp_get_digestbyname(name)
+      unless md
+        oid = LibCrypto.obj_txt2obj(name, 0)
+        md = LibCrypto.evp_get_digestbyname(LibCrypto.obj_nid2sn(LibCrypto.obj_obj2nid(oid)))
+        LibCrypto.asn1_object_free(oid)
+      end
+      unless md
+        raise "Unsupported digest algoritm: #{name}"
+      end
+      ctx = LibCrypto.evp_md_ctx_create
+      unless ctx
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      if LibCrypto.evp_digestinit_ex(ctx, md, nil) != 1
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      ctx
+    end
+
+    def self.new(name)
+      new(name, create_evp_mt_ctx(name))
```

Comment:
Isn't it trying to load a digest by oid when loading by name failed in the outlined part? I don't know if it's any useful, thought.